File: PRACTICAS/practica3/Civitas_P1/src/civitas_p1/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author angel
 */
public class CivitasJuego {
    
    private int indiceJugadorActual;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private EstadosJuego estado;
    private GestorEstados gestorEstados;
    private ArrayList<Jugador> jugadores;
    
    
    public CivitasJuego(ArrayList<String> nombres){
        
        jugadores = new ArrayList<>();
        
        for (int i = 0; i < nombres.size(); i++){
            Jugador player = new Jugador(nombres.get(i));
            jugadores.add(player);
        }
        
        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();
        indiceJugadorActual = Dado.instance.quienEmpieza(jugadores.size());
        tablero = new Tablero(1);
        mazo = new MazoSorpresas();
        inicializaTablero(mazo);
        inicializaMazoSorpresas(tablero);
        
    }
    
    
    private void inicializaTablero(MazoSorpresas mazo){
     
        tablero = new Tablero(8);
        
        TituloPropiedad propiedad1 = new TituloPropiedad("ESPAÑA", 400, (float) 1.2, 711, 1492, 600);
        Casilla casilla1 = new Casilla(propiedad1);
        tablero.añadeCasilla(casilla1);
        
        TituloPropiedad propiedad2 = new TituloPropiedad("ITALIA", 450, (float) 1.2, 752, 1523, 650);
        Casilla casilla2 = new Casilla(propiedad2);
        tablero.añadeCasilla(casilla2);
        
        TituloPropiedad propiedad3 = new TituloPropiedad("ALEMANIA", 700, (float) 1.2, 900, 1800, 900);
        Casilla casilla3 = new Casilla(propiedad3);
        tablero.añadeCasilla(casilla3);
        
        Casilla s1 = new Casilla(mazo, "SORPRESA UNO");
        tablero.añadeCasilla(s1);
        
        TituloPropiedad propiedad4 = new TituloPropiedad("BELGICA", 500, (float) 1.2, 810, 1570, 700);
        Casilla casilla4 = new Casilla(propiedad4);
        tablero.añadeCasilla(casilla4);
        
        TituloPropiedad propiedad5 = new TituloPropiedad("RUSIA", 440, (float) 1.2, 748, 1510, 635);
        Casilla casilla5 = new Casilla(propiedad5);
        tablero.añadeCasilla(casilla5);
        
        TituloPropiedad propiedad6 = new TituloPropiedad("EEUU", 720, (float) 1.2, 930, 1850, 925);
        Casilla casilla6 = new Casilla(propiedad6);
        tablero.añadeCasilla(casilla6);
        
        TituloPropiedad propiedad7 = new TituloPropiedad("JAPON", 325, (float) 1.2, 610, 1320, 510);
        Casilla casilla7 = new Casilla(propiedad7);
        tablero.añadeCasilla(casilla7);
        
        TituloPropiedad propiedad8 = new TituloPropiedad("COREA_DEL_NORTE", 243, (float) 1.2, 523, 1200, 401);
        Casilla casilla8 = new Casilla(propiedad8);
        tablero.añadeCasilla(casilla8);
        
        Casilla s2 = new Casilla(mazo, "SORPRESA DOS");
        tablero.añadeCasilla(s2);
        
        TituloPropiedad propiedad9 = new TituloPropiedad("GRECIA", 200, (float) 1.2, 490, 1150, 350);
        Casilla casilla9 = new Casilla(propiedad9);
        tablero.añadeCasilla(casilla9);
        
        TituloPropiedad propiedad10 = new TituloPropiedad("MADAGASCAR", 350, (float) 1.2, 620, 1330, 520);
        Casilla casilla10 = new Casilla(propiedad10);
        tablero.añadeCasilla(casilla10);
        
        Casilla parking = new Casilla("PARKING");
        tablero.añadeCasilla(parking);
        
        tablero.añadeJuez();
        
        TituloPropiedad propiedad11 = new TituloPropiedad("EL CARIBE", 750, (float) 1.2, 970, 1200, 960);
        Casilla casilla11 = new Casilla(propiedad11);
        tablero.añadeCasilla(casilla11);
        
        Casilla s3 = new Casilla(mazo, "SORPRESA TRES");
        tablero.añadeCasilla(s3);
        
        Casilla impuesto = new Casilla((float)800.0, "IMPUESTO");
        tablero.añadeCasilla(impuesto);
        
        TituloPropiedad propiedad12 = new TituloPropiedad("CHILE", 450, (float) 1.2, 758, 1200, 642);
        Casilla casilla12 = new Casilla(propiedad12);
        tablero.añadeCasilla(casilla12);
        
    }
    
    private void inicializaMazoSorpresas(Tablero tablero){
        
        Sorpresa ir_carcel = new Sorpresa(tablero, TipoSorpresa.IRCARCEL);
        Sorpresa ir_casilla = new Sorpresa(tablero, TipoSorpresa.IRCASILLA, 6, "Ve a la casilla numero 6");
        Sorpresa ir_casilla2 = new Sorpresa(tablero, TipoSorpresa.IRCASILLA, 13, "Ve a la casilla numero 13");
        Sorpresa librarte_carcel = new Sorpresa(TipoSorpresa.SALIRCARCEL, mazo);
        Sorpresa paga = new Sorpresa(TipoSorpresa.PAGARCOBRAR, -150, "Pagas 150" );
        Sorpresa cobra = new Sorpresa(TipoSorpresa.PAGARCOBRAR , 150, "Recibes 150");
        Sorpresa paga_edificio = new Sorpresa(TipoSorpresa.PORCASAHOTEL, -40, "Costes por casas y hoteles");
        Sorpresa cobra_edifico = new Sorpresa(TipoSorpresa.PORCASAHOTEL, 40, "Beneficios por casas y hoteles");
        Sorpresa pagar_a_jugador = new Sorpresa( TipoSorpresa.PORJUGADOR, -20, "Pagas 20 a cada jugador");
        Sorpresa cobrar_de_jugador = new Sorpresa(TipoSorpresa.PORJUGADOR, 20, "Recibes 20 de cada jugador");
        mazo.alMazo(ir_carcel);
        mazo.alMazo(ir_casilla);
        mazo.alMazo(ir_casilla2);
        mazo.alMazo(librarte_carcel);
        mazo.alMazo(paga);
        mazo.alMazo(cobra);
        mazo.alMazo(paga_edificio);
        mazo.alMazo(cobra_edifico);
        mazo.alMazo(pagar_a_jugador);
        mazo.alMazo(cobrar_de_jugador);

    }
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){

        while (tablero.getPorSalida() > 0){
            jugadorActual.pasaPorSalida();
        }
        
    }
    
    private void pasarTurno(){
        
        indiceJugadorActual = (indiceJugadorActual+1)%jugadores.size();
        
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        estado = gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), estado, operacion);
    }
    
    public boolean construirCasa(int ip){
        return jugadores.get(indiceJugadorActual).construirCasa(ip);
    }
    
    public boolean construirHotel(int ip){
        return jugadores.get(indiceJugadorActual).construirHotel(ip);
    }
    
    
    
    public boolean hipotecar(int ip){
        return jugadores.get(indiceJugadorActual).hipotecar(ip);
    }
    
    public boolean cancelarHipoteca(int ip){
        return jugadores.get(indiceJugadorActual).cancelarHipoteca(ip);
    }
  
    public boolean comprar(){
        
        Jugador jugadorActual=jugadores.get(indiceJugadorActual);
        int numCasillaActual= jugadorActual.getNumCasillaActual();
        Casilla casilla=tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo = casilla.getTituloPropiedad();
        boolean res = jugadorActual.comprar(titulo);
        return res;
    
    }
    
    public OperacionesJuego siguientePaso(){
    
        Jugador jugadorActual= jugadores.get(indiceJugadorActual);
        OperacionesJuego operacion= gestorEstados.operacionesPermitidas(jugadorActual, estado);
        if(operacion==OperacionesJuego.PASAR_TURNO){
            
            pasarTurno();
            siguientePasoCompletado(operacion);
        
        }else if(operacion==OperacionesJuego.AVANZAR){
            
            avanzaJugador();
            siguientePasoCompletado(operacion);
            
        }
        
        return operacion;
    }

    
    public boolean salirCarcelPagando(){
        return jugadores.get(indiceJugadorActual).salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        return jugadores.get(indiceJugadorActual).salirCarcelTirando();
    }
    
    public boolean finalDelJuego(){
        
       for (int i = 0; i < jugadores.size(); i++){
           if(jugadores.get(i).enBancarrota())
               return true;
       }
       
       return false;
        
    }
    
    public boolean vender(int ip){
       return jugadores.get(indiceJugadorActual).vender(ip);
    }
    
    public ArrayList<Jugador> ranking(){
        
        Collections.sort(jugadores);
        
        System.out.println("\nRanking de Jugadores!!!\n");
        
        for (int i = 0; i < jugadores.size();i++){
            System.out.println(jugadores.get(i).toString() + "\n");
        }
        
        return jugadores;
        
    }
        
    public Casilla getCasillaActual(){
        return tablero.getCasilla(jugadores.get(indiceJugadorActual).getNumCasillaActual());
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    public String infoJugadorTexto(){
        
        Jugador player = new Jugador(jugadores.get(indiceJugadorActual));
        return player.toString();
        
    }  
    /*
    private void avanzaJugador(){
        
        Jugador jugadorActual;
        int posicionActual;
        int tirada;
        int posicionNueva;
        Casilla casilla;
        
        jugadorActual = jugadores.get(indiceJugadorActual);
        
        posicionActual = jugadores.get(indiceJugadorActual).getNumCasillaActual();
        
        tirada = Dado.getInstance().tirar();
        
        
        posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
        
        casilla = tablero.getCasilla(posicionNueva);
        
        contabilizarPasosPorSalida(jugadorActual); BIEN
        jugadorActual.moverACasilla(posicionNueva); BIEN
        
        casilla.recibeJugador(indiceJugadorActual, jugadores);
        
        contabilizarPasosPorSalida(jugadorActual);
        
        
        
    }
    */
    private void avanzaJugador(){
        
        int posicionActual = jugadores.get(indiceJugadorActual).getNumCasillaActual();
        
        int tirada = Dado.getInstance().tirar();
        
        int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
        
        Casilla casilla = tablero.getCasilla(posicionNueva);
        
        contabilizarPasosPorSalida(jugadores.get(indiceJugadorActual));
        jugadores.get(indiceJugadorActual).moverACasilla(posicionNueva);
        
        casilla.recibeJugador(indiceJugadorActual, jugadores);
        
        contabilizarPasosPorSalida(jugadores.get(indiceJugadorActual));
        
        
        
    }
    
    
        
        
        
    }
        

    

File: PRACTICAS/practica3/otroruby/lib/civitas_juego.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative ('jugador')

module Civitas
class CivitasJuego

  
  def initialize(nombres)
    
    @gestor_estados = Gestor_estados.new
    @jugadores = Array.new
    @tablero = Tablero.new(3)
    @estado = Estados_juego::INICIO_TURNO
    @indice_jugador_actual = Dado.instance.quien_empieza(nombres.length)
    @mazo = MazoSorpresas.new
    
    for i in (0..nombres.size-1)
            
      nuevo_jugador = Jugador.jugador1(nombres.at(i))
      @jugadores.push(nuevo_jugador)
            
    end
   

    inicializar_mazo_sorpresas(@tablero)
    inicializar_tablero(@mazo)
    
  
    
  end
  
  private 
  
  def avanza_jugador

        jugador_actual = @jugadores[@indice_jugador_actual]

        posicion_actual = @jugadores[@indice_jugador_actual].num_casilla_actual

        tirada = Dado.instance.tirar


        posicion_nueva = @tablero.nueva_posicion(posicion_actual, tirada)

        casilla = @tablero.get_casilla(posicion_nueva)

        contabilizar_pasos_por_salida(jugador_actual)
        jugador_actual.mover_a_casilla(posicion_nueva)

        casilla.recibe_jugador(@indice_jugador_actual, @jugadores)

        contabilizar_pasos_por_salida(jugador_actual)

  end

  public
  
  def cancelar_hipoteca(ip)
    return @jugadores.at(@indice_jugador_actual).cancelar_hipoteca(ip);
  end
  
  def comprar
    res = false
    jugador_actual=@jugadores.get(@indice_jugador_actual)
    num_casilla_act= jugador_actual.num_casilla_actual
    casilla=@tablero.get_casilla(num_casilla_act)
    titulo = casilla.get_titulo_propiedad
    res = jugador_actual.comprar(titulo)
    
    return res
    
  end
  
  def construir_casa(ip)
    return @jugadores.at(@indice_jugador_actual).construir_casa(ip)
  end
  
  def construir_hotel(ip)
    return @jugadores.at(@indice_jugador_actual).construir_hotel(ip)
  end
  
  private
  
  def contabilizar_pasos_por_salida(jugador_actual)
        
    while(@tablero.get_por_salida>0)
      jugador_actual.pasa_por_salida
    end
  end
  
  public
  
  def final_del_juego
        
    for i in (0..@jugadores.size-1)
      
      if @jugadores.at(i).en_bancarrota
        
        return true        
      end
    end
    return false
  end
  
  def get_casilla_actual      
       return @tablero.get_casilla(@jugadores.at(@indice_jugador_actual).num_casilla_actual)
  end

  def get_jugador_actual      
        return @jugadores[@indice_jugador_actual]
  end
  
  def hipotecar(ip)
    return @jugadores.get(@indice_jugador_actual).hipotecar(ip);
  end

  def info_jugador_texto
        
    #nuevo = Jugador.new(@jugadores[@indice_jugador_actual])
    #info = "Nombre del jugador: #{nuevo.nombre} 
    #        Esta en la casilla: #{get_casilla_actual} 
    #        Tiene #{nuevo.propiedades.size} propiedades 
    #        Saldo del jugador: #{nuevo.saldo}" 
        
    #if(nuevo.is_encarcelado)
    #  info += " Esta encarcelado" 
    #else
    #  info += " No esta encarcelado" 
    #
    #end
  
    #return info
     
    return @jugadores.at(@indice_jugador_actual).to_s
    
  end      
  
  private
  
  def inicializar_mazo_sorpresas(tablero)
        
    ir_carcel = Sorpresa.new(TipoSorpresa::IRCARCEL, tablero)
    
    ir_casilla = Sorpresa.new(TipoSorpresa::IRCASILLA, tablero, 6, "Ve a la casilla numero 6")
    ir_casilla2 = Sorpresa.new(TipoSorpresa::IRCASILLA, tablero, 13, "Ve a la casilla numero 13")
    
    librarte_carcel = Sorpresa.new(TipoSorpresa::SALIRCARCEL, @mazo)
    
    paga = Sorpresa.new(TipoSorpresa::PAGARCOBRAR, -150, "Pagas 150")
    cobra = Sorpresa.new(TipoSorpresa::PAGARCOBRAR, 150, "Recibes 150")
    
    paga_edificios = Sorpresa.new(TipoSorpresa::PORCASAHOTEL, -40, "Costes por casas y hoteles")
    cobra_edificios = Sorpresa.new(TipoSorpresa::PORCASAHOTEL, 40, "Beneficios por casas y hoteles")
    
    pagar_a_jugador = Sorpresa.new(TipoSorpresa::PORJUGADOR, -20, "Pagas 20 a cada jugador")
    cobrar_de_jugador = Sorpresa.new(TipoSorpresa::PORJUGADOR, 20, "Recibes 20 de cada jugador")
    
    
    @mazo.al_mazo(ir_carcel)
    @mazo.al_mazo(ir_casilla)
    @mazo.al_mazo(ir_casilla2)
    @mazo.al_mazo(librarte_carcel)
    @mazo.al_mazo(paga)
    @mazo.al_mazo(cobra)
    @mazo.al_mazo(paga_edificios)
    @mazo.al_mazo(cobra_edificios)
    @mazo.al_mazo(pagar_a_jugador)
    @mazo.al_mazo(cobrar_de_jugador)
    
    
  end
    
    
    
  def inicializar_tablero(mazo)
        
    @tablero = Tablero.new(8)
   
    
    propiedad1 = TituloPropiedad.new("Espania",400,1.2,711,1492,600)
    casilla1 = Casilla.casilla(propiedad1)
    @tablero.aniade_casilla(casilla1)
    
    propiedad2 = TituloPropiedad.new("Italia",450,1.2,752,1523,650)
    casilla2 = Casilla.casilla(propiedad2)
    @tablero.aniade_casilla(casilla2)
    
    propiedad3 = TituloPropiedad.new("Alemania",700,1.2,900,1800,900)
    casilla3 = Casilla.casilla(propiedad3)
    @tablero.aniade_casilla(casilla3)
    
    propiedad4 = TituloPropiedad.new("Belgica",500,1.2,810,1570,700)
    casilla4 = Casilla.casilla(propiedad4)
    @tablero.aniade_casilla(casilla4)
    
    propiedad5 = TituloPropiedad.new("Rusia",440,1.2,748,1510,635)
    casilla5 = Casilla.casilla(propiedad5)
    @tablero.aniade_casilla(casilla5)
    
    propiedad6 = TituloPropiedad.new("EEUU",720,1.2,930,1850,925)
    casilla6 = Casilla.casilla(propiedad6)
    @tablero.aniade_casilla(casilla6)
    
    propiedad7 = TituloPropiedad.new("Japon",325,1.2,610,1320,510)
    casilla7 = Casilla.casilla(propiedad7)
    @tablero.aniade_casilla(casilla7)
    
    propiedad8 = TituloPropiedad.new("Corea del Norte",243,1.2,523,1200,401)
    casilla8 = Casilla.casilla(propiedad8)
    @tablero.aniade_casilla(casilla8)
    
    propiedad9 = TituloPropiedad.new("Grecia",200,1.2,490,1150,350)
    casilla9 = Casilla.casilla(propiedad9)
    @tablero.aniade_casilla(casilla9)
    
    propiedad10 = TituloPropiedad.new("Madagascar",350,1.2,620,1330,520)
    casilla10 = Casilla.casilla(propiedad10)
    @tablero.aniade_casilla(casilla10)
    
    
    parking = Casilla.new("parking")
    @tablero.aniade_casilla(parking)
    
    
    propiedad11 = TituloPropiedad.new("El Caribe",750,1.2,970,1200,960)
    casilla11 = Casilla.casilla(propiedad11)
    @tablero.aniade_casilla(casilla11)
    
    impuesto = Casilla.new(1000,"impuesto")
    @tablero.aniade_casilla(impuesto)
    
    propiedad12 = TituloPropiedad.new("Chile",450,1.2,758,1200,642)
    casilla12 = Casilla.casilla(propiedad12)
    @tablero.aniade_casilla(casilla12)
    
    sorpresa1 = Casilla.new(mazo, "sorpresa1")
    @tablero.aniade_casilla(sorpresa1)
    
    @tablero.aniade_juez
    
    sorpresa2 = Casilla.new(mazo, "sorpresa2")
    @tablero.aniade_casilla(sorpresa2)
    
    sorpresa3 = Casilla.new(mazo, "sorpresa3")
    @tablero.aniade_casilla(sorpresa3)
  end
  
  def pasar_turno      
        @indice_jugador_actual = (@indice_jugador_actual +1)%@jugadores.size       
  end
 
  def ranking
        
    rank = Array.new
    copia_original = Array.new
        
    for i in (0..@jugadores.size-1)
      copia_original.push(@jugadores[i])
    end  
        
    contador = 0 
        
    while (contador < @jugadores.size)
            
      mayor = 0;
            
      for j in (1..copia_original.size-1)
                
        if(copia_original[mayor].<=>(copia_original[j]) == -1)
          mayor = j               
        end
      end
            
      rank.push(@jugadores[mayor]);
      @jugadores.delete(@jugadores[mayor]);
            
      contador+=1
            
    end
        
        return rank
  end
  
  public

  def salir_carcel_pagando
     
    return @jugadores[@indice_jugador_actual].salir_carcel_pagando
  end
    
  def salir_carcel_tirando
    
    return @jugadores[@indice_jugador_actual].salir_carcel_tirando
  end

  def siguiente_paso_completado(operacion)
     
    @estado = @gestor_estados.siguiente_estado(@jugadores[@indice_jugador_actual], @estado, operacion)
        
  end
    
  def vender(ip)
        
    return @jugadores[@indice_jugador_actual].vender(ip)
        
  end
  def siguiente_paso
    
        jugador_actual= @jugadores.at(@indice_jugador_actual)
        operacion= @gestor_estados.operaciones_permitidas(jugador_actual, @estado)
        if operacion==Operaciones_Juego::PASAR_TURNO
            
            pasar_turno
            siguiente_paso_completado(operacion)
        
        elsif operacion==Operaciones_Juego::AVANZAR
            
            avanza_jugador
            siguiente_paso_completado(operacion)
            
        end
        
        return operacion
  end


end
end